mlir/lib/Dialect/Rock/IR/GetRockInfo.cpp

```diff
@@ -0,0 +1,140 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Rock/IR/GetRockInfo.h"
+
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Rock/IR/AmdArchDb.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Value.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/ErrorHandling.h"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+FailureOr<StringAttr> mlir::rock::getArch(Operation *op) {
+  return getAttrFromOpOrParents<StringAttr>(op, "arch", "mhal.arch");
+}
+
+StringAttr mlir::rock::getArchValue(Operation *op) {
+  auto maybeArch = rock::getArch(op);
+  if (failed(maybeArch))
+    llvm_unreachable("No 'arch' attribute on kernel");
+
+  return maybeArch.value();
+}
+
+FailureOr<int64_t> mlir::rock::getNumCU(Operation *op) {
+  FailureOr<StringAttr> maybeArch = getArch(op);
+  if (failed(maybeArch)) {
+    return failure();
+  }
+  StringAttr arch = maybeArch.value();
+  FailureOr<IntegerAttr> maybeNumCU =
+      getAttrFromOpOrParents<IntegerAttr>(op, "num_cu", "numCU");
+  if (failed(maybeNumCU)) {
+    return failure();
+  }
+  IntegerAttr numCU = maybeNumCU.value();
+  AmdArchInfo archInfo = rock::lookupArchInfo(arch);
+  if (numCU.getValue().getSExtValue() < archInfo.minNumCU) {
+    return op->emitError() << "num_cu=" << numCU
+                           << " cannot be lower than arch minNumCU="
+                           << archInfo.minNumCU;
+  }
+  return numCU.getValue().getSExtValue();
+}
+
+int64_t mlir::rock::getNumCUValue(Operation *op) {
+  auto maybeCU = rock::getNumCU(op);
+  if (succeeded(maybeCU)) {
+    return maybeCU.value();
+  }
+
+  // Otherwise, we will need to get the minimum CU value from the architecture
+  auto archStr = rock::getArchValue(op);
+  int64_t minCU = rock::lookupArchInfo(archStr).minNumCU;
+  return minCU;
+}
+
+bool opHasOptionalFeature(Operation *op) {
+  bool hasOptionalFeature = llvm::TypeSwitch<Operation*,
+                                             bool>(op)
+  .Case<rock::GemmOp, rock::ConvOp, rock::ConvBwdDataOp,
+        rock::ConvBwdWeightOp,  rock::GridwiseGemmOp,
+        rock::GridwiseGemmAccelOp, rock::AttentionOp, rock::ReduceOp,
+        rock::GlobalStoreOp, rock::GemmElementwiseGemmOp,
+        rock::ConvElementwiseGemmOp, rock::ThreadwiseWriteAllOp,
+        rock::BlockwiseGemmAccelOp, rock::ThreadwiseAccelGemmOp,
+        rock::ConvertingCopyKernelOp,
+        rock::GridwiseAttentionAccelOp>([](auto opWithFeatures) {
+
+    return true;
+  })
+  .Default([](Operation *op) -> bool {
+    return false;
+  });
+  
+  return hasOptionalFeature;
+}
+
+GemmFeatures intersectGemmFeatures(GemmFeatures a, GemmFeatures b) {
+  return static_cast<GemmFeatures>(static_cast<uint32_t>(a) &
+                                   static_cast<uint32_t>(b));
```

Comment:
You can just do return `a & b` 

TableGen generates it automatically 
See definition of GemmFeatures 
```
inline constexpr GemmFeatures operator&(GemmFeatures a, GemmFeatures b) {
  return static_cast<GemmFeatures>(static_cast<uint32_t>(a) & static_cast<uint32_t>(b));
}
```